src/Compilers/Core/Portable/PEWriter/PeWriter.cs

```diff
@@ -88,10 +89,17 @@ internal sealed class PeWriter
             // If PDB writer is given, we have to have PDB path.
             Debug.Assert(nativePdbWriterOpt == null || pdbPathOpt != null);
 
-            var peWriter = new PeWriter(context.Module.Properties, context.Module.Win32Resources, context.Module.Win32ResourceSection, pdbPathOpt, deterministic);
-            var mdWriter = FullMetadataWriter.Create(context, messageProvider, allowMissingMethodBodies, deterministic, getPortablePdbStreamOpt != null, cancellationToken);
+            try
+            {
+                var peWriter = new PeWriter(context.Module.Properties, context.Module.Win32Resources, context.Module.Win32ResourceSection, pdbPathOpt, deterministic);
+                var mdWriter = FullMetadataWriter.Create(context, messageProvider, allowMissingMethodBodies, deterministic, getPortablePdbStreamOpt != null, cancellationToken);
 
-            return peWriter.WritePeToStream(mdWriter, getPeStream, getPortablePdbStreamOpt, nativePdbWriterOpt);
+                return peWriter.WritePeToStream(mdWriter, getPeStream, getPortablePdbStreamOpt, nativePdbWriterOpt);
+            }
+            catch (Exception ex) when (!(ex is PdbWritingException || ex is ResourceException || ex is PermissionSetFileReadException))
```

Comment:
All those catches in PdbWriter handle exceptions are thrown in the interop stub by COM marshaller one frame lower. They never catch any exceptions that might originate from our code or from user code.

Re catch around StrongNameProvider.CreateInputStream - it covers only small amount of our code. It would indeed be marginally better if this exception handling was moved into the body of CreateInputStream. 

Most other cases that I can find cover only small amount of code. My problem with this change is that this catch covers the entire implementation of PE writer, metadata writer, PDB writer and user code.